File: common/src/main/java/com/ssblur/alchimiae/events/network/client/ReceiveIngredientsNetwork.java
package com.ssblur.alchimiae.events.network.client;

import com.ssblur.alchimiae.AlchimiaeMod;
import com.ssblur.alchimiae.alchemy.ClientAlchemyHelper;
import com.ssblur.alchimiae.events.network.AlchimiaNetworkInterface;
import dev.architectury.networking.NetworkManager;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;

import java.util.List;

public class ReceiveIngredientsNetwork implements AlchimiaNetworkInterface<ReceiveIngredientsNetwork.Payload> {
  @Override
  public CustomPacketPayload.Type<Payload> type() {
    return Payload.TYPE;
  }

  @Override
  public StreamCodec<RegistryFriendlyByteBuf, Payload> streamCodec() {
    return Payload.STREAM_CODEC;
  }

  @Override
  public NetworkManager.Side side() {
    return NetworkManager.Side.S2C;
  }

  @Override
  public void receive(Payload value, NetworkManager.PacketContext context) {
    AlchimiaeMod.LOGGER.debug("Received data for ingredient with key {}", value.key);
    ClientAlchemyHelper.update(value.key, value.effects);
  }

  public record Payload(String key, List<String> effects) implements CustomPacketPayload {
    public static final Type<Payload> TYPE = new Type<>(AlchimiaeMod.location("receive_ingredients"));
    public static final StreamCodec<RegistryFriendlyByteBuf, Payload> STREAM_CODEC = StreamCodec.composite(
      ByteBufCodecs.STRING_UTF8,
      Payload::key,
      ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
      Payload::effects,
      Payload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
      return TYPE;
    }
  }
}

File: common/src/main/kotlin/com/ssblur/alchimiae/events/network/client/ReceiveIngredientsNetwork.kt
package com.ssblur.alchimiae.events.network.client

import com.ssblur.alchimiae.AlchimiaeMod
import com.ssblur.alchimiae.alchemy.ClientAlchemyHelper
import com.ssblur.alchimiae.events.network.AlchimiaNetworkInterface
import dev.architectury.networking.NetworkManager
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload

class ReceiveIngredientsNetwork : AlchimiaNetworkInterface<ReceiveIngredientsNetwork.Payload> {
  override fun type(): CustomPacketPayload.Type<Payload> {
    return Payload.TYPE
  }

  override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf?, Payload> {
    return Payload.STREAM_CODEC
  }

  override fun side(): NetworkManager.Side {
    return NetworkManager.Side.S2C
  }

  override fun receive(value: Payload, context: NetworkManager.PacketContext) {
    AlchimiaeMod.LOGGER.debug("Received data for ingredient with key {}", value.key)
    ClientAlchemyHelper.update(value.key, value.effects)
  }

  data class Payload(val key: String, val effects: List<String>) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
      return TYPE
    }

    companion object {
      val TYPE: CustomPacketPayload.Type<Payload> =
        CustomPacketPayload.Type(AlchimiaeMod.location("receive_ingredients"))
      val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf?, Payload> = StreamCodec.composite(
        ByteBufCodecs.STRING_UTF8,
        Payload::key,
        ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
        Payload::effects
      ) { key: String, effects -> Payload(key, effects) }
    }
  }
}
